packages/grafana-data/src/transformations/transformers/calculateField.ts

```diff
@@ -1,82 +1,261 @@
-import { DataFrame, DataTransformerInfo, Vector, FieldType, Field, NullValueMode } from '../../types';
+import { DataFrame, DataTransformerInfo, Vector, FieldType, Field, NullValueMode, FieldConfig } from '../../types';
 import { DataTransformerID } from './ids';
 import { ReducerID, fieldReducers } from '../fieldReducer';
 import { getFieldMatcher } from '../matchers';
 import { FieldMatcherID } from '../matchers/ids';
 import { RowVector } from '../../vector/RowVector';
-import { ArrayVector } from '../../vector';
+import { ArrayVector, BinaryOperationVector, ConstantVector } from '../../vector';
 import { doStandardCalcs } from '../fieldReducer';
+import { seriesToColumnsTransformer } from './seriesToColumns';
+import { getTimeField } from '../../dataframe';
+import defaults from 'lodash/defaults';
+import { BinaryOperationID, binaryOperators } from '../../utils/binaryOperators';
 
-export interface CalculateFieldTransformerOptions {
-  reducer: ReducerID;
+export enum CalculateFieldMode {
+  ReduceRow = 'reduceRow',
+  BinaryOperation = 'binary',
+  Scale = 'scale',
+}
+
+interface ReduceOptions {
   include?: string; // Assume all fields
-  alias?: string; // The output field name
-  replaceFields?: boolean;
+  reducer: ReducerID;
   nullValueMode?: NullValueMode;
 }
 
+interface BinaryOptions {
+  left: string;
+  operator: BinaryOperationID;
+  right: string;
+}
+
+interface ScaleOptions {
+  left: string;
+  operator: BinaryOperationID;
+  right: number;
+}
+
+const defaultReduceOptions: ReduceOptions = {
+  reducer: ReducerID.sum,
+};
+
+const defaultBinaryOptions: BinaryOptions = {
+  left: '',
+  operator: BinaryOperationID.Add,
+  right: '',
+};
+
+const defaultScaleOptions: ScaleOptions = {
+  left: '',
+  operator: BinaryOperationID.Multiply,
+  right: 1,
+};
+
+export interface CalculateFieldTransformerOptions {
+  // True/False or auto
+  timeSeries?: boolean;
+  mode: CalculateFieldMode; // defaults to 'reduce'
+
+  // Only one should be filled
+  reduce?: ReduceOptions;
+  binary?: BinaryOptions;
+  scale?: ScaleOptions;
+
+  // Remove other fields
+  replaceFields?: boolean;
+
+  // Output field properties
+  alias?: string; // The output field name
+  config?: FieldConfig;
+}
+
+/**
+ * Return true if every frame has a time column
+ */
+function findTimeSeriesName(data: DataFrame[]): string | undefined {
```

Comment:
The description is a bit off, returns true if frame has time column, but return type is string | undefined. 

The name is a bit confusing as well. getConsistentTimeFieldName ?

Reply:
ha!  guess what the first version of the function did?  but since the `join` transformer we have works on the name, this had to find the name field and can not just assume "time"

Reply:
yes, but it does not return the name of the time series but the time field.